timemachine/fe/mcgregor.py

```diff
@@ -135,45 +126,105 @@ def __init__(self, n_vertices, edges):
+        # args = (mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size)
+        # assert self.mapping_is_disconnected_fast(*args) == self.mapping_is_disconnected_ref(*args)
+
+        return self.mapping_is_disconnected_fast(
+            mapped_nodes, unvisited_nodes, max_connected_components, min_connected_component_size
+        )
+
+    def mapping_is_disconnected_fast(
+        self,
+        mapped_nodes: Set[int],
+        unvisited_nodes: Set[int],
+        max_connected_components: Optional[int],
+        min_connected_component_size: int,
+    ):
+        """Optimized implementation of mapping_is_disconnected, sacrificing modularity for efficiency"""
+
+        seen = set()
+        n_ccs = 0
+        for u in mapped_nodes:
+            if u not in seen:
+                # visit component containing v
+                seen.add(u)
+                cc_size = 1
+                nextlevel = [u]
+                while nextlevel:
+                    thislevel = nextlevel
+                    nextlevel = []
+                    for v in thislevel:
+                        for w in self.get_neighbors(v):
+                            if w in mapped_nodes or w in unvisited_nodes:
+                                if w not in seen:
+                                    seen.add(w)
+                                    cc_size += 1
+                                    nextlevel.append(w)
+                n_ccs += 1
+                if cc_size < min_connected_component_size:
+                    return True
+                if n_ccs == max_connected_components:
+                    # if we've seen the maximum number of connected components, we should have seen all of the mapped nodes
+                    return not mapped_nodes.issubset(seen)
+
+        return False
+
+    def mapping_is_disconnected_ref(
+        self,
+        mapped_nodes: Set[int],
+        unvisited_nodes: Set[int],
+        max_connected_components: Optional[int],
+        min_connected_component_size: int,
+    ):
+        """Reference implementation of mapping_is_disconnected, decomposed into standard algorithms for clarity"""
+
+        g = nx.Graph(self.edges)
+
+        # Consider the subgraph induced by mapped and unvisited nodes (ignoring visited nodes that have not been mapped)
+        sg = g.subgraph(mapped_nodes | unvisited_nodes)
+
+        n_ccs_with_mapped_nodes = 0
+
+        for cc in nx.connected_components(sg):
+            if cc.intersection(mapped_nodes):
+                n_ccs_with_mapped_nodes += 1
+
+                # The number of connected components in the subgraph is a lower bound on the eventual number (since we could
+                # choose not to map nodes that are currently unvisited). Note that we do not revisit nodes that are visited
+                # but have not been mapped, so the number of connected components will not shrink.
```

Comment:
Clarification, does this sentence just mean:

Visiting the remaining nodes can only increase or maintain the number of connected components.

Reply:
Yes, that is a much more succinct summary. Will improve the comment.

Reply:
Improved comments in https://github.com/proteneer/timemachine/commit/9f0bac8cc167008dffa7fdad2fc0e12909a7d4cc